js/src/components/tree-select-control/index.js

```diff
@@ -1,56 +1,190 @@
-	const { className, disabled, value = [], onChange = () => {} } = props;
-	const [ isExpanded, setIsExpanded ] = useState( true );
+/**
+ * The Option type Object. This is how we send the options to the selector.
+ *
+ * @typedef { Object } Option
+ * @property {string} id The unique ID for the option
+ * @property {string} name The name for the option
+ * @property {Option[]} [children] The children Option objects
+ */
+
+/**
+ * Renders a component with a searchable control, tags and a tree selector.
+ *
+ * @param {Object} props Component props.
+ * @param {string} props.id Component id
+ * @param {string} props.label Label for the component
+ * @param {string} props.placeholder Placeholder for the search control input
+ * @param {string} props.className The class name for this component
+ * @param {boolean} props.disabled Disables the component
+ * @param {Option[]} props.options Options to show in the component
+ * @param {{string}[]} props.value Selected values
+ * @param {Function} props.onChange Callback when the selector changes
+ *
+ * @return {JSX.Element|null} The component
+ */
+const TreeSelectControl = ( {
+	id,
+	label,
+	placeholder,
+	className,
+	disabled,
+	options = [],
+	value = [],
+	onChange = () => {},
+} ) => {
+	const [ isExpanded, setIsExpanded ] = useState( false );
+	const focusOutside = useFocusOutside( () => {
+		setIsExpanded( false );
+	} );
+
+	/**
+	 * Optimizes the performance for getting the tags info
+	 *
+	 * @see getTags
+	 */
+	const optionsRepository = useMemo( () => {
+		const repository = {};
+		options.forEach( ( option ) => {
+			option.children.forEach( ( child ) => {
+				repository[ child.id ] = { ...child, parent: option.id };
+			} );
+		} );
+
+		return repository;
+	}, [ options ] );
+
+	if ( ! options.length ) {
+		return null;
+	}
+
+	/**
+	 * Get formatted Tags from the selected values.
+	 *
+	 * @return {{name: {string}, id: {string}}[]} An array of Tags
+	 */
+	const getTags = () => {
+		return value.map( ( key ) => {
+			const option = optionsRepository[ key ];
+			return { id: key, name: option.name };
+		} );
+	};
+
+	/**
+	 * Handles a change on the Tree List of options. Could be a click on a parent option
+	 * or a child option
+	 *
+	 * @param {boolean} checked Indicates if the item should be checked
+	 * @param {Option} option The option to change
+	 */
+	const handleListChange = ( checked, option ) => {
+		if ( option.children?.length ) {
+			handleParentChange( checked, option );
+		} else {
+			handleSingleChange( checked, option );
+		}
+	};
+
+	/**
+	 * Handles a change of a child element.
+	 *
+	 * @param {boolean} checked Indicates if the item should be checked
+	 * @param {Option} option The option to change
+	 */
+	const handleSingleChange = ( checked, option ) => {
+		const idPosition = value.indexOf( option.id );
+		const newValue = [ ...value ];
+
+		if ( ! checked ) {
+			newValue.splice( idPosition, 1 );
+		} else {
+			newValue.push( option.id );
+		}
```

Comment:
💅  More succinct 😄  :

```suggestion
		const newValue = checked
			? [ ...value, option.id ]
			: value.filter( ( el ) => el !== option.id );
```